### PR TITLE
Image-rs & CDH | Refactoring and use the same ImageClient

### PR DESCRIPTION
**Pull request by Xynnn007**
This PR organizes the code of image-rs to make it more modular and allows CDH and image-rs to share the same image pull configuration file.

With the modular refactoring, we found some hidden bugs that caused by static/global variables. This refactoring tries to get rid of static/global variables to promote the reentrancy.

Please see each commit for a more structured view.

cc @fitzthum @ChengyuZhu6

### CODE REVIEW

File: image-rs/src/signature/mod.rs
Comment (fitzthum) [inline image-rs/src/signature/mod.rs:25]: This is kind of an abstract name for a struct. Maybe it could be something that reflects the config like `ImageSecurityConfig` or something that reflects what it is doing like `ImageSecurityChecker` It seems like this mainly for signatures so maybe `SignatureValidator` or something could make sense.

File: image-rs/src/signature/mod.rs
Comment (fitzthum) [inline image-rs/src/signature/mod.rs:55]: maybe `check_image_requirement` and `check_cosign_image_requirement` would be more clear?

File: image-rs/src/signature/mod.rs
Comment (fitzthum) [inline image-rs/src/signature/mod.rs:77]: `check_image_signature`?

File: image-rs/src/builder.rs
Comment (fitzthum) [inline image-rs/src/builder.rs:94]: also quite a vague name. if we change the name of `ImageSecurity` we can adopt something similar here.

File: image-rs/src/builder.rs
Comment (fitzthum) [inline image-rs/src/builder.rs:106]: Maybe we should add a warning if this isn't set?

File: image-rs/src/resource/kbs/mod.rs
Comment (fitzthum) [inline image-rs/src/resource/kbs/mod.rs:55]: nit: add "sample" or "default" to this message to make it slightly easier to find

File: image-rs/src/resource/mod.rs
Comment (fitzthum) [inline image-rs/src/resource/mod.rs:43]: I wonder if we should let people specify a URI that starts with `file://`

File: image-rs/src/signature/policy/mod.rs
Comment (fitzthum) [inline image-rs/src/signature/policy/mod.rs:46]: We might want to keep something like this warning in our replacement method.

**Review by fitzthum — COMMENTED** [review]
A big improvement imo. I left a few comments.

File: image-rs/src/builder.rs
Comment (Xynnn007) [reply image-rs/src/builder.rs:94]: How about `image_security_module`?

File: image-rs/src/builder.rs
Comment (Xynnn007) [reply image-rs/src/builder.rs:106]: Good idea

File: image-rs/src/image.rs
Comment (Xynnn007) [reply image-rs/src/image.rs:197]: Good catch. Combined. Thanks!

File: image-rs/src/resource/kbs/mod.rs
Comment (Xynnn007) [reply image-rs/src/resource/kbs/mod.rs:55]: Fixed, thanks!

File: image-rs/src/signature/mod.rs
Comment (Xynnn007) [reply image-rs/src/signature/mod.rs:25]: `ImageSignatureValidator` sounds good.

File: image-rs/src/signature/mod.rs
Comment (Xynnn007) [reply image-rs/src/signature/mod.rs:55]: `check_image_requirement` is good.

File: image-rs/src/signature/mod.rs
Comment (Xynnn007) [reply image-rs/src/signature/mod.rs:77]: nice suggestion. Fixed.

File: image-rs/src/signature/policy/mod.rs
Comment (Xynnn007) [reply image-rs/src/signature/policy/mod.rs:46]: Added above `check_image_signature()`

**Review by fitzthum — APPROVED** [review]
LGTM

I think there are still some improvements to make in image-rs, but this is a step forward.

**Review by ChengyuZhu6 — COMMENTED** [review]
Thanks for the work @Xynnn007. Just a comment:

**Comment by Xynnn007** [comment]
> I think there are still some improvements to make in image-rs, but this is a step forward.

Yes. Please give some ideas about this, and I will try to figure out if we can promote in this pr together

File: image-rs/tests/image_decryption.rs
Comment (Xynnn007) [reply image-rs/tests/image_decryption.rs:47]: Yes.

Comment (Xynnn007) [reply image-rs/tests/image_decryption.rs:47]: btw The `ClientBuilder` logic brings `image policy`, `auth file`, `sigstore config` fetching in the time of creation of `ClientBuilder.build()`, and supports to read them from local filesystem.

**Comment by Xynnn007** [comment]
Rebased to resolve the conflicts.

File: image-rs/tests/image_decryption.rs
Comment (ChengyuZhu6) [reply image-rs/tests/image_decryption.rs:47]: Got it. You mean that `ClientBuilder.build()` would read some configs by default that are irrelevant to image descryption?

**Review by ChengyuZhu6 — APPROVED** [review]
LGTM thanks for the work @Xynnn007. BTW,  I have a question that’s unrelated to this PR:
I’ve noticed that some comments in the code of the repo use three slashes `///`, while others use two `//`. Are there specific scenarios for using these two types of comments? For example, is it recommended to use `///` in one comment type  and `//`  in another comment type?

**Comment by fitzthum** [comment]
> I've noticed that some comments in the code of the repo use three slashes ///, while others use two //. Are there specific scenarios for using these two types of comments? For example, is it recommended to use /// in one comment type and // in another comment type?

Three slash comments get used to generate the rustdocs. You can also use `/** */` to do block doc comments. Ideally we should have one of these comments at the top of all our public functions in components that can be used as crates.

**Comment by ChengyuZhu6** [comment]
> > I've noticed that some comments in the code of the repo use three slashes ///, while others use two //. Are there specific scenarios for using these two types of comments? For example, is it recommended to use /// in one comment type and // in another comment type?
> 
> Three slash comments get used to generate the rustdocs. You can also use `/** */` to do block doc comments. Ideally we should have one of these comments at the top of all our public functions in components that can be used as crates.

Thanks Tobin. I noticed that some public structures or variables in the structure in the code are commented with `//`, while others are commented with `///`. This seems curious to me.

**Comment by ChengyuZhu6** [comment]
BTW, do we need to cut a minor release and bump guest-components and image-rs to the new minor release in kata, according to the community meeting? If so, I recommend holding off on merging this PR until the release is cut.

File: image-rs/tests/image_decryption.rs
Comment (Xynnn007) [reply image-rs/tests/image_decryption.rs:47]: Oh. I mean that `image policy`, `auth file`, `sigstore config` all are explicitly set in the config file. If any/all of them is set, the read will be performed by `ClientBuilder.build()`

**Comment by Xynnn007** [comment]
Added another commit to replace HTTPS_PROXY and NO_PROXY env setting with config to `oci-client`.

**Comment by Xynnn007** [comment]
Added another commit to support self-signed image registry certs in CDH/image-rs config.

Seeing that cosign will also need to fetch signatures from the same registry, I made a PR to support this in sigstore-rs crate. https://github.com/sigstore/sigstore-rs/pull/392

After that PR is merged we can make the story complete

**Comment by Xynnn007** [comment]
Add one more commit for cosign to support fetch signature self-signed registry and proxy. I think it is now the PR is ready.

**Pull request merged**
